analyzers/src/SonarAnalyzer.CSharp/Rules/AssertionArgsShouldBePassedInCorrectOrder.cs

```diff
@@ -18,63 +18,50 @@
-                {
-                    return;
-                }
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class AssertionArgsShouldBePassedInCorrectOrder : SonarDiagnosticAnalyzer
+{
+    internal const string DiagnosticId = "S3415";
+    private const string MessageFormat = "Make sure these 2 arguments are in the correct order: expected value, actual value.";
 
-                var methodCallExpression = (MemberAccessExpressionSyntax)methodCall.Expression;
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
 
-                var methodKnownTypes = MethodsWithType.GetValueOrDefault(methodCallExpression.Name.Identifier.ValueText);
-                if (methodKnownTypes == null)
-                {
-                    return;
-                }
+    private static readonly IDictionary<string, ImmutableArray<KnownType>> MethodsWithType = new Dictionary<string, ImmutableArray<KnownType>>
+    {
+        ["AreEqual"]    = ImmutableArray.Create(KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_Assert, KnownType.NUnit_Framework_Assert),
+        ["AreNotEqual"] = ImmutableArray.Create(KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_Assert, KnownType.NUnit_Framework_Assert),
+        ["AreSame"]     = ImmutableArray.Create(KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_Assert, KnownType.NUnit_Framework_Assert),
+        ["AreNotSame"]  = ImmutableArray.Create(KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_Assert, KnownType.NUnit_Framework_Assert),
+        ["Equal"]       = ImmutableArray.Create(KnownType.Xunit_Assert),
+        ["Same"]        = ImmutableArray.Create(KnownType.Xunit_Assert),
+        ["NotSame"]     = ImmutableArray.Create(KnownType.Xunit_Assert)
+    };
 
-                var symbolInfo = c.SemanticModel.GetSymbolInfo(methodCallExpression.Expression).Symbol;
-                var isAnyTrackedAssertType = (symbolInfo as INamedTypeSymbol).IsAny(methodKnownTypes);
-                if (!isAnyTrackedAssertType)
-                {
-                    return;
-                }
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(c =>
+        {
+            if (c.Node is InvocationExpressionSyntax { ArgumentList: { Arguments: { Count: >= 2 } arguments } argumentList, Expression: MemberAccessExpressionSyntax methodCallExpression }
+                && methodCallExpression.IsKind(SyntaxKind.SimpleMemberAccessExpression)
+                && CSharpFacade.Instance.MethodParameterLookup(argumentList, c.SemanticModel) is var parameterLookup
+                && CheckArguments(parameterLookup)
+                && MethodsWithType.GetValueOrDefault(methodCallExpression.Name.Identifier.ValueText) is { } methodKnownTypes
```

Comment:
You may want to try to turn the `MethodsWithType` dictionary into a switch expression instead. It might be better readable than this lookup after all.
Such a switch expression could be a private method and you might also pass in the "Container of the invocation" (Container = The INamedTypeSymbol "class Assert" of the IMethodSymbol of the invocation) and just return true/false.
This would also allow the check for the "unexpected" name if we expand the switch even more.